chips/rp2040/src/uart.rs

```diff
@@ -0,0 +1,733 @@
+enum UARTStateRX {
+    Idle,
+    Receiving,
+    AbortRequested,
+}
+
+const UART0_BASE: StaticRef<UartRegisters> =
+    unsafe { StaticRef::new(0x40034000 as *const UartRegisters) };
+
+const UART1_BASE: StaticRef<UartRegisters> =
+    unsafe { StaticRef::new(0x40038000 as *const UartRegisters) };
+
+pub struct Uart<'a> {
+    registers: StaticRef<UartRegisters>,
+    interrupt: u32,
+    tx_client: OptionalCell<&'a dyn TransmitClient>,
+    rx_client: OptionalCell<&'a dyn ReceiveClient>,
+
+    tx_buffer: TakeCell<'static, [u8]>,
+    tx_position: Cell<usize>,
+    tx_len: Cell<usize>,
+    tx_status: Cell<UARTStateTX>,
+
+    rx_buffer: TakeCell<'static, [u8]>,
+    rx_position: Cell<usize>,
+    rx_len: Cell<usize>,
+    rx_status: Cell<UARTStateRX>,
+}
+
+impl<'a> Uart<'a> {
+    pub const fn new_uart0() -> Self {
+        Self {
+            registers: UART0_BASE,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+
+            tx_buffer: TakeCell::empty(),
+            tx_position: Cell::new(0),
+            tx_len: Cell::new(0),
+            tx_status: Cell::new(UARTStateTX::Idle),
+
+            rx_buffer: TakeCell::empty(),
+            rx_position: Cell::new(0),
+            rx_len: Cell::new(0),
+            rx_status: Cell::new(UARTStateRX::Idle),
+            interrupt: UART0_IRQ,
+        }
+    }
+    pub const fn new_uart1() -> Self {
+        Self {
+            registers: UART1_BASE,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+
+            tx_buffer: TakeCell::empty(),
+            tx_position: Cell::new(0),
+            tx_len: Cell::new(0),
+            tx_status: Cell::new(UARTStateTX::Idle),
+            rx_buffer: TakeCell::empty(),
+            rx_position: Cell::new(0),
+            rx_len: Cell::new(0),
+            rx_status: Cell::new(UARTStateRX::Idle),
+            interrupt: UART1_IRQ,
+        }
+    }
+
+    pub fn enable(&self) {
+        self.registers.uartcr.modify(UARTCR::UARTEN::SET);
+    }
+
+    pub fn disable(&self) {
+        self.registers.uartcr.modify(UARTCR::UARTEN::CLEAR);
+    }
+
+    pub fn enable_transmit_interrupt(&self) {
+        self.registers.uartimsc.modify(UARTIMSC::TXIM::SET);
+    }
+
+    pub fn disable_transmit_interrupt(&self) {
+        self.registers.uartimsc.modify(UARTIMSC::TXIM::CLEAR);
+    }
+
+    pub fn enable_receive_interrupt(&self) {
+        self.registers.uartifls.modify(UARTIFLS::RXIFLSEL::FIFO_1_8);
+
+        self.registers.uartimsc.modify(UARTIMSC::RXIM::SET);
+    }
+
+    pub fn disable_receive_interrupt(&self) {
+        self.registers.uartimsc.modify(UARTIMSC::RXIM::CLEAR);
+    }
+
+    fn uart_is_writable(&self) -> bool {
+        !self.registers.uartfr.is_set(UARTFR::TXFF)
+    }
+
+    pub fn send_byte(&self, data: u8) {
+        while !self.uart_is_writable() {}
+        self.registers.uartdr.write(UARTDR::DATA.val(data as u32));
+    }
```

Comment:
This type of blocking interface is not favored in the tock kernel, see https://github.com/tock/tock/blob/master/doc/Design.md#split-phase-operation . While some exceptions exist, they are generally only for cases where blocking is expected to be very short

Reply:
This is used only for panic

Reply:
Oh, thanks, sorry I didn't trace through and realize that. In that case it is fine.